stdlib/std.jsonnet

```diff
@@ -88,6 +88,41 @@ limitations under the License.
     range(from, to)::
         std.makeArray(to - from + 1, function(i) i + from),
 
+    slice(indexable, index, end, step)::
+        if (index != null && index < 0) || (end != null && end < 0) || (step != null && step < 0) then
+            error("got [%s:%s:%s] but negative index, end, and steps are not implemented" % [index, end, step] )
+        else if std.type(indexable) != "string" && std.type(indexable) != "array" then
+            error("std.slice accepts a string or an array, but got: %s" % std.type(indexable))
+        else
+            local invar =
+                // loop invariant with defaults applied
+                {
+                    indexable: indexable,
+                    index:
+                        if index == null then 0
+                        else index,
+                    end:
+                        if end == null then std.length(indexable)
+                        else end,
+                    step:
+                        if step == null then 1
+                        else step,
+                    length: std.length(indexable),
+                    type: std.type(indexable)
+                };
+            local build(slice, cur) =
+                if cur >= invar.end || cur >= invar.length then
+                    slice
+                else
+                    build(
+                        if invar.type == "string" then
+                            slice + invar.indexable[cur]
+                        else
+                            slice + [ invar.indexable[cur] ],
+                        cur + invar.step
+                    ) tailstrict;
```

Comment:
Yes I am not sure how to solve this.  There are various instances of it throughout the code.

The Python approach is to make std.join O(n) and then do everything in terms of that.

Currently std.join is O(n^2) and making it O(n) would mean making it native.

There may be a better way though.